### ISSUE TITLE
macOS Home Assistant Companion App: No Accessibility Features

### ISSUE THREAD

**Issue by misterbond007** [home-assistant/iOS#3327]
**Device model, version and app version**

Model Name: Mac mini 2024 Apple M4
macOS Version: macOS Sequoia 15.2
App Version: Home Assistant Companion App version 2024.12.2 (2024.1058)
Safari Version: Version 18.2 (20620.1.16.11.8)

**Home Assistant Core Version**

2025.1.2

**Describe the bug**
Home Assistant does not correctly report the focused UI element using Apple's Accessibility features. When a text field in Home Assistance has keyboard focus, the Accessibility API does not report a UI element with the role “text field” or “text area” as having focus, but rather a generic group. This prevents Accessibility-enabled tools from being able to detect when a user might need assistance typing into a text field.

Specifically, the issue is that, when a text field is focused in Home Assistant, the "AXFocusedUIElement" attribute of the system-wide UI element does not indicate a UI element with the role “AXTextField” or “AXTextArea” is focused.

**To Reproduce**
Install macOS 15.2
Install Home Assistant Companion App version 2024.12.2 (2024.1058)
On a touch screen set the focus into a text field

**Expected behavior**
When a text field in Home Assistance Companion has keyboard focus, the Accessibility API should report a UI element with the role “text field” or “text area” as having focus.

**Comment by bgoncal** [comment]
Thanks for reporting the issue, currently the App only displays a webview with Home Assistant frontend inside, so I'd recommend reporting this in the [frontend repo](https://github.com/home-assistant/frontend) since the App is not in control on what displays in the webview.

**Comment by misterbond007** [comment]
I have reported it in the frontend repo, but they said, I should post it here :-)

**Comment by bgoncal** [comment]
Could you link me to the frontend issue?

**Comment by misterbond007** [comment]
https://github.com/home-assistant/frontend/issues/23664

**Comment by misterbond007** [comment]
Don't you think the issue is because of the webview, that the macOS accesibilty feature only sees a "group", but not a text field?

**Comment by bgoncal** [comment]
You need to compare to Safari to evaluate that, but in the App can't you focus on anything inside the webview? a button for example.

**Comment by misterbond007** [comment]
I have tested it now in Safari too, without any problems: the accesibilty feature works (the text field is recognized as a text field). I can focus on buttons and on everything. But as described, the accessibility feature doesn't recognize the items correctly.

**Comment by bgoncal** [comment]
Ok, then this will need further investigation since it looks like a bug on the webview (WKWebView) itself, the behavior should be the same as Safari.
Please let me know all differences you have noticed between safari and the app and when I have time I'll investigate this.

**Comment by misterbond007** [comment]
Ok, thank you.
As I know, the accesibility feature sees everything as a "Group", but not for example as the correct specific item (e.g. text field).

The issue is that, when a text field is focused in Home Assistant, the "AXFocusedUIElement" attribute of the system-wide UI element does not indicate a UI element with the role “AXTextField” or “AXTextArea” is focused.

**Comment by misterbond007** [comment]
How far are the investigations on this?

**Comment by bgoncal** [comment]
It looks like a bug on Apple's side for webviews in Mac Catalyst apps, till this moment I haven't found a workaround for it
